### PR TITLE
keyborg:focusout event and the fixes for Shadow DOM handling.

### PR DESCRIPTION
**Pull request by mshoho**
Adding `keyborg:focusout` custom event that works properly in the Shadow DOM scenarios.
Fixing how the Shadow Root focus listeners are attached and removed.
Fixing a case when Keyborg is created when something in a shadow is focused to make sure this shadow gets focus listeners attached.

### CODE REVIEW

File: src/FocusEvent.ts [hunk 1]
```diff
@@ -152,28 +199,32 @@ export function setupFocusEvent(win: Window): void {
...
        *         > focused element ✅ (no shadow root - dispatch keyborg event)
        */
-      target.shadowRoot.addEventListener("focusin", focusInHandler, true);
-      target.shadowRoot.addEventListener(
-        "focusout",
-        focusOutShadowRootHandler,
-        true,
-      );
+
+      if (shadowTargets.has(shadowRoot)) {
```
Comment (ling1726) [inline src/FocusEvent.ts:213]: Let's add a test to make sure the edge case isn't forgotten?

Comment (mshoho) [reply src/FocusEvent.ts:216]: I have added a test to check if the event handlers are attached, reattached and removed properly.

File: tests/shadow-dom.stories.tsx [hunk 2]
```diff
@@ -47,3 +47,72 @@ export const Nested = () => (
     <KeyboardMode />
   </div>
 );
+
+interface WindowWithCreateKeyborg extends Window {
+  createKeyborg?: typeof createKeyborg;
+}
+
+export const Lazy = () => {
+  (window as WindowWithCreateKeyborg).createKeyborg = createKeyborg;
```
Comment (ling1726) [inline tests/shadow-dom.stories.tsx:56]: instead of using an eval to create keyborg from a global instance, add a button that creates an instance of keyborg on click

Comment (ling1726) [reply tests/shadow-dom.stories.tsx:56]: Likewise the event listeners can be added on the example side and output to the DOM.

File: src/FocusEvent.ts [hunk 3]
```diff
@@ -152,28 +202,32 @@ export function setupFocusEvent(win: Window): void {
...
+        return;
+      }
+
+      shadowRoot.addEventListener("focusin", focusInHandler, true);
+      shadowRoot.addEventListener("focusout", focusOutHandler, true);
+
+      shadowTargets.add(shadowRoot);
```
Comment (ling1726) [inline src/FocusEvent.ts:223]: if we're going to be keeping shadowRoot references - let's make them weakrefs

File: src/FocusEvent.ts [hunk 4]
```diff
@@ -221,6 +283,16 @@ export function setupFocusEvent(win: Window): void {
     return origFocus.apply(this, arguments);
   }
 
+  let activeElement = kwin.document.activeElement as Element | null;
+
+  // If keyborg is created with the focus inside shadow root, we need
+  // to go through the shadows up to make sure all relevant shadows
+  // have focus handlers attached.
+  while (activeElement && activeElement.shadowRoot) {
```
Comment (ling1726) [inline src/FocusEvent.ts:291]: do we need to drill down to all shadow roots? if so what about parent shadow roots? do we need to handle them all too?

**Review by ling1726 — APPROVED** [review]

File: tests/shadow-dom.stories.tsx (hunk 2 above)
Comment (mshoho) [reply tests/shadow-dom.stories.tsx:56]: I'll do a follow-up on how the tests are written.
